[C++ verification] New issue with ESBMC for C++: std::isnan

This issue was reported by one of our industrial partners:

````
After I updated the repo, ESBMC started failing while compiling the code. I was able to reproduce the failure on a small test.

 
Here is the content of the cpp file:
````

````CPP
#include <cstdint>
#include <cmath>


    float max(float a, float b)

    {

        assert(!std::isnan(a) && "Max<float> A was NaN!");

        assert(!std::isnan(b) && "Max<float> B was NaN!");

        return ((a > b) ? a : b);

    }


int main()

{

    float n1, n2;

 

    return max(n1, n2);

}
````
 
````
The command is here: esbmc main.cpp -Wno-writable-strings --std c++11 --verbosity clang:9 "-Wc,--gcc-install-dir=/nfs/site/itools/em64t_SLES12SP5/pkgs/gcc/12.1.0/lib/gcc/x86_64-pc-linux-gnu/12.1.0".
````
 

Here is the failure:

````
main.cpp:6:22: error: expected unqualified-id

        assert(!std::isnan(a) && "Max<float> A was NaN!");

                     ^

/usr/include/math.h:268:6: note: expanded from macro 'isnan'

     (sizeof (x) == sizeof (float)                                            \

     ^

main.cpp:7:22: error: expected unqualified-id

        assert(!std::isnan(b) && "Max<float> B was NaN!");

                     ^

/usr/include/math.h:268:6: note: expanded from macro 'isnan'

     (sizeof (x) == sizeof (float)                                            \

     ^

ERROR: PARSING ERROR
````

--- a/regression/esbmc-cpp/bug_fixes/github_1998/main.cpp
+++ b/regression/esbmc-cpp/bug_fixes/github_1998/main.cpp
@@ -1,0 +1,17 @@
+#include <cmath>
+#include <cassert>
+
+float max(float a, float b)
+
+{
+  assert(std::isnan(a) && "Max<float> A was NaN!");
+
+  assert(std::isnan(b) && "Max<float> B was NaN!");
+
+  return ((a > b) ? a : b);
+}
+int main()
+{
+  float n1 = NAN, n2 = NAN;
+  return max(n1, n2);
+}

--- a/regression/esbmc-cpp/bug_fixes/github_1998/test.desc
+++ b/regression/esbmc-cpp/bug_fixes/github_1998/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.cpp
+
+^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-cpp/bug_fixes/github_1998_fail/main.cpp
+++ b/regression/esbmc-cpp/bug_fixes/github_1998_fail/main.cpp
@@ -1,0 +1,17 @@
+#include <cmath>
+#include <cassert>
+
+float max(float a, float b)
+
+{
+  assert(!std::isnan(a) && "Max<float> A was NaN!");
+
+  assert(!std::isnan(b) && "Max<float> B was NaN!");
+
+  return ((a > b) ? a : b);
+}
+int main()
+{
+  float n1, n2;
+  return max(n1, n2);
+}

--- a/regression/esbmc-cpp/bug_fixes/github_1998_fail/test.desc
+++ b/regression/esbmc-cpp/bug_fixes/github_1998_fail/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.cpp
+
+^VERIFICATION FAILED$

--- a/src/cpp/library/cmath
+++ b/src/cpp/library/cmath
@@ -149,6 +149,8 @@ template <typename Integer>
 double modf(Integer num, double *iptr);
  */
 
+using ::__builtin_isnan;
+
 #undef CIMPORT1
 #undef CIMPORT2
 

